### DISCUSSION TITLE
Database's

### DISCUSSION THREAD

**Discussion by dubiouscript** [TiddlyWiki/TiddlyWiki5#7347] [Ideas]
reading about this  https://github.com/Jermolene/TiddlyWiki5/pull/7329 - demo of alternate wiki store implementation 
 pull request and saw mention of sqlite 

which reminded meh of read about  [eatherpad](https://github.com/ether/etherpad-lite) using  https://github.com/ether/ueberDB 

>UeberDB turns every database into a simple key value store by providing a layer of abstraction between your software and your database.

>Database Support
>     SQLite
>    Couch, Dirty, Elasticsearch, Maria
>    memory: An in-memory ephemeral database.
>   Mongo, MsSQL, MySQL,  Postgres (single connection and with connection pool)
>   Redis, Rethink

"turning every database into a simple key value store" is (imho) an interesting idea
i wander if  such an idea  can provide any inspiration wrt  "future experiments with wiki stores" :thinking: 

---

also worth a mention i think is this comparison of relational data tools
[Hyperpolyglot : Relational Data Tools: SQL, Awk, Pig](https://hyperpolyglot.org/data)  - a side-by-side reference sheet

which make me wander about *exactly* why (what criteria) ? ... lead to the *speculative* focus on sql ? :thinking: 

any way , :eyes:  im elated :star_struck:  to read of about the motivation / aspiration  for these experiments !

>The plan is to build it up into the smallest possible plain JS wiki store implementation, sharing as much implementation as possible with the existing core implementation with as little code duplication as possible. 

:sunglasses: :+1: :+1: :+1:

**Comment by dubiouscript** [comment]
ot/rant 

while reading/thinking about "future experiments with wiki stores" and the goal
- smallest possible plain JS wiki store implementation

i was reminded of previously working on a few things ( both js)

0) one of the  FUSE bindings for Javascript and node.js  implementations ( https://github.com/mafintosh/fuse-bindings or perhaps https://github.com/bcle/fuse4js ) 
where i added some (at the time) missing functionality stat and sym-links afair (never pushed/published ftr)
 
the FUSE bindings
can be (simply) described as 
 "maping" between the filesystem api and json object(s) 

1) and writing a ( verrrry basic (barely functional)) www cache / url router  (writen/hosted ) on https://github.com/dvbportal/appjet

the cache/router
can be simply described as 
"mapping" between the `origin:www-url(s)` and  the `host/app:cached-urls`

both operate in very different domains   file system vs web , 
but had (imho) much similarity wrt resolving requested resources ( via some array/map )  

while trying to pick up tiddly wiki ( reading a bit of source ect)

i get a similar sense  / impression  with TW that (somewhere) in the core of tiddly wiki's  "storage" and "retrieval" function 

is the same/similar "mapping" occurring 

tbh im not shore if these observations are either coherent (to my self or others) or relevant (to tw generally)

but it dose appear that tw ( depending on svr config or single page mode) 
 operate by  "mapping" between both file system vs web

what i mean is 
i can view a file system like tree ( eg exploring the core ) 
from a single html file  mapping between , tid (json) to html "virtual" html file system view 

(and) or (with server) map between filesystem (plain text files file:///abc/xyz.txt ect) to   a single html file 

*mind-boggles* 

point being that  perhaps some where at the core of the "smallest possible plain JS wiki store implementation" is 
a kind of (plain js) mapped / index / array / lookup :thinking: 

/rant 

also this comment appears relevant and clarify's the sql implementaion under consideration is a `WASM version of SQLite` :open_mouth: 

https://github.com/Jermolene/TiddlyWiki5/discussions/7282#discussioncomment-5110011 # TW be like Graph Databases

---
>Hi @daStretchyCat it is fair to view TiddlyWiki as special kind of graph database. We even use indexes to optimise queries just like a real database.
>
>The core was designed to allow other database engines to be dropped in, but there hasn't been a great deal of work on it.
>I'm particularly interested in using the WASM version of SQLite. 
>The only requirement is that the core database must have a synchronous API. Databases that have an asynchronous API must instead be integrated via a syncer.
---

>just like a real database.

reminded me of this link : http://freesoftwaremagazine.com/articles/look_mum_no_database_thanks_to_awk/
:grinning: :stuck_out_tongue_winking_eye: 
see also the Hyperpolyglot : Relational Data Tools: SQL, Awk, Pig - a side-by-side reference sheet
link in first post :sunglasses: 
+
https://github.com/dbohdan/sqawk # Like awk but with SQL and table joins :monocle_face:

**Comment by linonetwo** [comment]
About SQLite implementation https://github.com/Jermolene/TiddlyWiki5/discussions/7531

**Reply by dubiouscript** [reply]
+ ftr

https://pql.dev/ - pql is an open-source pipelined query language that translates to SQL :dark_sunglasses: 

where pipelines intertwingle with sequences

**Reply by Jermolene** [reply]
Thanks @dubiouscript that is very interesting. The ppl language has some similarities to our filter language. It is also interesting to see the compiled SQL.

**Reply by dubiouscript** [reply]
@Jermolene no problem ! , ftr there are a few other similar projects which can (afaik) compile to SQL 
what are the similarities you see with tiddly wiki filters ?
...my guess would be command pipeline!
